Pattern Fingerprint Issues

**Pattern fingerprints no longer work as expected:**

<!-- Please provide information about your setup -->
<!-- To get the RDKit version use the following commands:
import rdkit
rdkit.__version__
-->
- RDKit Version: Don't know how to get C++ version/Pull of master at 5pm MST 6/17/2019
- Platform: Linux 64bit

<!-- Please describe your problem as detailed as necessary -->

Pattern fingerprints for certain queries are not working as expected.  There are cases where the AllProbeBitsMatch function returns false and the query is a substructure of the target.  These patterns all used to work.

<!-- If applicable, please provide code that can reproduce your issue -->
```

#include <iostream>
#include <GraphMol/GraphMol.h>
#include <GraphMol/SmilesParse/SmilesParse.h>
#include <GraphMol/Fingerprints/Fingerprints.h>
#include <DataStructs/ExplicitBitVect.h>
#include <DataStructs/BitOps.h>
#include <GraphMol/Substruct/SubstructMatch.h>

void testFingerprint(const std::string &smarts, const std::string &smiles) {
    auto target = RDKit::SmilesToMol(smiles);
    auto query = RDKit::SmartsToMol(smarts);
    auto targetFp = RDKit::PatternFingerprintMol(*target, 2048);
    auto queryFp = RDKit::PatternFingerprintMol(*query, 2048);
    auto match = AllProbeBitsMatch(*queryFp, *targetFp);
    delete targetFp;
    delete queryFp;
    std::cout << "Pattern fingerprint from " << smarts << (match ? " matches" : " does not match") << " target "
              << smiles << std::endl;
    RDKit::MatchVectType matchVect;
    auto isSubstructure = RDKit::SubstructMatch(*target, *query, matchVect);
    std::cout << "Pattern " << smarts << (isSubstructure ? " is a substructure of " : " is not a substructure of ")
              << " target " << smiles << std::endl << std::endl;
}

int main() {
    testFingerprint("*", "CN=C1SC=C(C)N1N=CC1=CC=CC=N1");
    testFingerprint("[#6D3](-*)(:*):[#7,#8]", "CN=C1SC=C(C)N1N=CC1=CC=CC=N1");
    testFingerprint("[ND2](*)=*", "CN=C1SC=C(C)N1N=CC1=CC=CC=N1");
    testFingerprint("[#7;D3]-*=*", "CN=C1SC=C(C)N1N=CC1=CC=CC=N1");
    testFingerprint("[#7;D3](*=*)(-&!@*)*:*", "CN=C1SC=C(C)N1N=CC1=CC=CC=N1");
}
```

On an older version of RDKit I get the expected output:
```
Pattern fingerprint from * matches target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern * is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [#6D3](-*)(:*):[#7,#8] matches target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [#6D3](-*)(:*):[#7,#8] is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [ND2](*)=* matches target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [ND2](*)=* is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [#7;D3]-*=* matches target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [#7;D3]-*=* is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [#7;D3](*=*)(-&!@*)*:* matches target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [#7;D3](*=*)(-&!@*)*:* is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
```

But now I see:

```
Pattern fingerprint from * does not match target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern * is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [#6D3](-*)(:*):[#7,#8] does not match target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [#6D3](-*)(:*):[#7,#8] is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [ND2](*)=* does not match target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [ND2](*)=* is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [#7;D3]-*=* does not match target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [#7;D3]-*=* is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1

Pattern fingerprint from [#7;D3](*=*)(-&!@*)*:* does not match target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
Pattern [#7;D3](*=*)(-&!@*)*:* is a substructure of  target CN=C1SC=C(C)N1N=CC1=CC=CC=N1
```
I discovered the issue running Java unit tests after rebuilding RDKit.  The tests run searches using all the smarts queries in the `patty_rules.txt` file.  The following patterns now fail (they all used to work):

```
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#6D3](=*)-[#7,#8] ==> expected: <378> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#6D3](:*)(:*)[#7,#8] ==> expected: <1378> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#6D3](-*)(:*):[#7,#8] ==> expected: <2000> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#6D2](*)#[#7,#8] ==> expected: <97> but was: <23>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [NH0]#* ==> expected: <97> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [ND2](*)=* ==> expected: <2000> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for * ==> expected: <2000> but was: <414>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7D1]-* ==> expected: <38> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7D2](-*)-* ==> expected: <218> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7D3](-*)(-*)-* ==> expected: <183> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7D1]-*=* ==> expected: <38> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7H;D2]-*=* ==> expected: <218> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7H;D2]-*:* ==> expected: <216> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7;D3]-*=* ==> expected: <2000> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7;D3]-*:* ==> expected: <381> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7;D3](*=*)(-&!@*)*:* ==> expected: <2000> but was: <0>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for [#7,#15,#33;D3]=* ==> expected: <151> but was: <14>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for *-[#8,#16;X1] ==> expected: <151> but was: <14>
[ERROR]   TestInMemorySubstructureSearch.testDatabaseSearch:38 Substructure search hits count error for O=C-*:[#6](*)[OH] ==> expected: <38> but was: <0>
```
Cheers,
Gareth

<!-- If you have a support question, please use the rdkit-discuss mailing list: 
https://sourceforge.net/p/rdkit/mailman/ -->

--- a/Code/GraphMol/Fingerprints/CMakeLists.txt
+++ b/Code/GraphMol/Fingerprints/CMakeLists.txt
@@ -29,5 +29,7 @@ rdkit_test(testFingerprints test1.cpp LINK_LIBRARIES
 rdkit_test(testFingerprintGenerators testFingerprintGenerators.cpp LINK_LIBRARIES 
            Fingerprints SmilesParse GraphMol DataStructs FileParsers
            RDGeneral )
+rdkit_catch_test(fpTestCatch catch_tests.cpp LINK_LIBRARIES Fingerprints SmilesParse GraphMol DataStructs FileParsers
+           RDGeneral)
 
 add_subdirectory(Wrap)

--- a/Code/GraphMol/Fingerprints/PatternFingerprints.cpp
+++ b/Code/GraphMol/Fingerprints/PatternFingerprints.cpp
@@ -198,12 +198,14 @@ ExplicitBitVect *PatternFingerprintMol(const ROMol &mol, unsigned int fpSize,
   boost::tie(firstA, lastA) = mol.getVertices();
   while (firstA != lastA) {
     const Atom *at = mol[*firstA];
-    if (isComplexQuery(at)) {
+    // isComplexQuery() no longer considers "AtomNull" to be complex, but for the purposes
+    // of the pattern FP, it definitely needs to be treated as a query feature.
+    if ( at->hasQuery() && (at->getQuery()->getDescription()=="AtomNull" || isComplexQuery(at))) {
       isQueryAtom.set(at->getIdx());
-      // std::cerr<<"   complex atom: "<<at->getIdx()<<std::endl;
     }
     ++firstA;
   }
+
   ROMol::EDGE_ITER firstB, lastB;
   boost::tie(firstB, lastB) = mol.getEdges();
   while (firstB != lastB) {
@@ -211,7 +213,6 @@ ExplicitBitVect *PatternFingerprintMol(const ROMol &mol, unsigned int fpSize,
     // if( isComplexQuery(bond) ){
     if (isPatternComplexQuery(bond)) {
       isQueryBond.set(bond->getIdx());
-      // std::cerr<<"   complex bond: "<<bond->getIdx()<<std::endl;
     }
     ++firstB;
   }

--- a/Code/GraphMol/Fingerprints/catch_tests.cpp
+++ b/Code/GraphMol/Fingerprints/catch_tests.cpp
@@ -1,0 +1,47 @@
+//
+//  Copyright (C) 2019 Greg Landrum
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#define CATCH_CONFIG_MAIN  // This tells Catch to provide a main() - only do
+                           // this in one cpp file
+#include "catch.hpp"
+
+#include <memory>
+#include <RDGeneral/test.h>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/FileParsers/FileParsers.h>
+#include <GraphMol/Fingerprints/Fingerprints.h>
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include <DataStructs/ExplicitBitVect.h>
+#include <DataStructs/BitOps.h>
+#include <string>
+
+using namespace RDKit;
+
+TEST_CASE("Github 2051", "[patternfp,bug]") {
+    auto mol="CCC1CC1"_smiles;
+    std::unique_ptr<ExplicitBitVect> mfp(PatternFingerprintMol(*mol));
+
+    REQUIRE(mfp);
+  SECTION("basics1") {
+      auto qmol="**"_smarts;
+      std::unique_ptr<ExplicitBitVect> qfp(PatternFingerprintMol(*qmol));
+      REQUIRE(qfp);
+
+      CHECK(AllProbeBitsMatch(*qfp,*mfp));   
+  }
+  SECTION("basics2") {
+      auto qmol="*"_smarts;
+      std::unique_ptr<ExplicitBitVect> qfp(PatternFingerprintMol(*qmol));
+      REQUIRE(qfp);
+      CHECK(AllProbeBitsMatch(*qfp,*mfp));
+  }
+}
+

